### DISCUSSION TITLE
Add toAsyncItetator support.

### DISCUSSION THREAD

**Discussion by SanderElias** [ReactiveX/rxjs#6332] [General]
## Feature Request

Add a `.toAsyncIterator()` method to rxjs. This will make the interop with es20018 better, and also adds an easy way to consume observables in a more readable form.
![image](https://user-images.githubusercontent.com/1249083/43939775-5e2a58ea-9c6c-11e8-863b-2d5a96566dcd.png)

**Describe the solution you'd like**
adding this to the prototype of rxjs, makes it easier to consume observables as a async iterator.
Here is the function I used in the above sample. 

```ts
async function* toAsyncIterator(ob: Observable<any>) {
  let goResolve, goReject;
  const setPromise = () =>
    new Promise<Res>((resolve, reject) => {
      goResolve = resolve;
      goReject = reject;
    });
  let lastPromise = setPromise();

  ob.subscribe(
    r =>
      goResolve({
        value: r,
        done: false
      }),
    err => goReject(err),
    () => goResolve({ value: undefined, done: true })
  );

  // Still have to decide what to to with the error case
  // for now, it will be forwarded to the consumer.
  // (they need to implement a try/catch)

  while (true) {
    const result: Res = await lastPromise;
    if (result.done) {
      return;
    }
    yield result.value;
    lastPromise = setPromise();
  }
}
```

**Describe alternatives you've considered**
The alternative is keep the function to myself. or perhaps add it to rxjs as an utility function. 

**Additional context**
Supporting async itterators makes it possible to consume observables in a more imparative way. That way it observables more aproachable for developers that have less of a functional background.

**Comment by cartant** [comment]
Related: the [notes from the last core-team meeting](https://github.com/ReactiveX/rxjs-core-notes/blob/master/rxjs/2018/2018-08/august-1.md) - see the last item.

**Comment by SanderElias** [comment]
@cartant The notes provided me with the idea for this. I should add an option that handles the buffer issue.
It needs to be an option because blindly buffering everything is probably a bad idea in most cases. Depending on the scenario it could need different handling option. 
Let me mull on this for a while. Perhaps it's more of a documentation issue. The buffer handling can be done before the 'subscribe/conversion'

**Comment by somombo** [comment]
@SanderElias I'm with you on allowing an imperative style for observables. I have a question though..

In your implementation `ob` is subscribed to... I understand that if `ob` errors or completes, it will automatically be unsubscribed. However since your `toAsyncIterator` makes a subscription but only returns an AsyncIterable, how can users of your proposed functionality manually get a hold of the `ob` subscription, in case they (for whatever reason) want to unsubscribe it at some point before `ob` errors or completes (e.g. for better resource management)??

**Comment by SanderElias** [comment]
@somombo Have a closer look at my sample. I'm using a takeUntil to do exactly what you are asking for.
If that's not what you meant, can you explain to me what it is you are looking for?

**Comment by somombo** [comment]
oh sorry I completely had misunderstood the roll of the `takeUntil` makes sense now. At least there is *some* way to destroy, at will if the `destroy` is a subject. Thanks for pointing it out.

**Comment by somombo** [comment]
okay so with this understanding in mind, I propose that the api for this functionality look something like this:

```typescript
interface Observable {
  /* ... */
  toAsyncIterator( destroyer?: Observable<any> )
  /* ... */
}
```

Note that the method `toAsyncIterator` has the `destroyer` as an optional parameter. If `destroyer` is provided then it will be incorporated *within* the body of `toAsyncIterator` using `takeUntil` just as u have exemplified  and if it is not provided then the method would be just as you currently have it. Essentially, the observable that's finally subscribed to will depend on whether `destroyer` is provided or not.

**Comment by alisahinozcelik** [comment]
Hi, what about adding `[Symbol.asyncIterator]` method directly to `Observable` instead of adding a new method.
Then the developers may just use observables in the for of loops like this:

```javascript
for await (const result of anObservable$) {
// Do some stuff
}
```

**Comment by ambientlight** [comment]
are there any plans landing this in 7.0.0 given that https://github.com/ReactiveX/rxjs/issues/1624 is there already?

**Comment by ambientlight** [comment]
@alisahinozcelik: how about something like this?

```ts
import { Observable, Subscriber } from "rxjs";
export interface AsyncIterableObservable<T> extends Observable<T>, AsyncIterable<T> {}

(Observable.prototype as AsyncIterableObservable<any>)[Symbol.asyncIterator] = async function*<T>() {
  let resolve: (value: { value: T | undefined, done: boolean } | PromiseLike<{ value: T | undefined, done: boolean}> | undefined) => void;
  let reject: ((reason: any) => void);

  const setPromise = () =>
    new Promise<{ value: T | undefined, done: boolean}>((_resolve, _reject) => {
      resolve = _resolve;
      reject = _reject;
    });
  let lastPromise = setPromise();

  const subscription = this.subscribe({
    next: emitted => { 
      resolve({ value: emitted, done: false })
      lastPromise = setPromise()
    },
    error: err => reject(err),
    complete: () => resolve({ value: undefined, done: true })
  });

  // Still have to decide what to to with the error case
  // for now, it will be forwarded to the consumer.
  // (they need to implement a try/catch)

  try {
    while (true) {
      const result = await lastPromise;
      if (result.done) {
        return;
      }

      yield result.value as T
    }
  } finally {
    subscription.unsubscribe()
  }
}
```

But please not this will not work for observables that resolve synchronously, so the next thing requires `observeOn(asyncScheduler)` to work:

```js
async function obsTest(){
  const obs = from([1, 2, 3, 4]).pipe(
    observeOn(asyncScheduler)) as AsyncIterableObservable<number>
  for await (const result of obs){
    console.log(result)
  }

  console.log("ended")
}
```

Also, moving `lastPromise = setPromise()` in OP's original code from generator loop to subscriber will result in `for await` resolving with no values rather then just 'freezing'.

I wonder how do you guys think this should be handled? It is supposed to buffer or something?

**Comment by ambientlight** [comment]
Had issues with teardown when the above is used in [subscriptions-transport-ws](https://github.com/apollographql/subscriptions-transport-ws),  revamped the above with imperative implementation:

```ts
import { Observable, Subscriber, Notification, NotificationKind } from "rxjs";
import { materialize } from "rxjs/operators";

export interface AsyncIterableObservable<T> extends Observable<T>, AsyncIterable<T> {}

(Observable.prototype as AsyncIterableObservable<any>)[Symbol.asyncIterator] = function<T>(){
  type RetainedPromiseFuncs = { 
    resolve: (value: IteratorResult<T>) => void, 
    reject: (reason?: any) => void
  };

  const pullQueue: RetainedPromiseFuncs[] = [];
  const pushQueue: Notification<T>[] = [];
  
  const resolveEmission = (promiseFnPair: RetainedPromiseFuncs, emission: Notification<T>) => {
    const { resolve, reject } = promiseFnPair;
    switch(emission.kind){
    case NotificationKind.NEXT: return resolve({ value: emission.value as T, done: false })
    case NotificationKind.ERROR: return reject(emission.error)
    case NotificationKind.COMPLETE: return resolve({ value: undefined, done: true })
    } 
  };

  const subscription = this.pipe(
    materialize()
  ).subscribe((emission: Notification<T>) => {
    const promiseFns = pullQueue.shift()
    if(promiseFns){
      resolveEmission(promiseFns, emission)
    } else {
      pushQueue.push(emission)
    }
  });

  const pull: () => Promise<IteratorResult<T>> = () => new Promise(
    (resolve, reject) => {
      const emission = pushQueue.shift()
      if(emission){
        resolveEmission({ resolve, reject }, emission)
      } else {
        pullQueue.push({ resolve, reject })
      }
    }
  )

  const teardown = () => {
    if(subscription.closed){
      return
    }

    subscription.unsubscribe()
    pullQueue.forEach(({ resolve, reject }) => resolve({ value: undefined, done: true }))
    pullQueue.length = 0
    pushQueue.length = 0
  }

  return {
    next(){
      return subscription.closed 
        ? (this.return as (() => Promise<IteratorReturnResult<void>>))() 
        : pull()
    },
    return() {
      teardown();
      return Promise.resolve({ value: undefined, done: true });
    },
    throw(error: any) {
      teardown();
      return Promise.reject(error)
    },
    [Symbol.asyncIterator]() {
      return this;
    }
  }
}
```

used originally with:

```ts
export function fromAsyncIterator<T>(input: AsyncIterable<T>): AsyncIterableObservable<T> {
  let ai = input[Symbol.asyncIterator]();
  const getNextValue: (subscriber: Subscriber<T>) => Promise<void> = subscriber => ai.next().then(
    result => {
      if(result.done){
        subscriber.complete()
      } else {
        subscriber.next(result.value)
        getNextValue(subscriber)
      }},
    err => subscriber.error(err));

  return new Observable(subscriber => { 
    getNextValue(subscriber);
    return () => ai.return ? ai.return() : undefined
  }) as AsyncIterableObservable<T>
}
```

Let me know if any thoughts.
